viewer/packages/camera-manager/src/Flexible/FlexibleCameraManager.ts

```diff
@@ -0,0 +1,551 @@
+      return;
+    }
+    // This is used to determine the speed of the camera when flying with ASDW.
+    // We want to either let it be controlled by the near plane if we are far away,
+    // but no more than a fraction of the bounding box of the system if inside
+    const diagonal = this.getBoundingBoxDiagonal2();
+    const diagonalFraction = diagonal * this.options.sensitivityDiagonalFraction;
+    const nearFraction = 0.1 * this.camera.near;
+
+    let sensitivity: number;
+    if (diagonalFraction < nearFraction) {
+      sensitivity = nearFraction;
+    } else {
+      sensitivity = this.options.getLegalSensitivity(diagonalFraction);
+    }
+    this.options.sensitivity = sensitivity;
+
+    // Nils: Need this for a while to fine tune
+    // console.log('diagonalFraction:     ', diagonalFraction);
+    // console.log('nearFraction:         ', nearFraction);
+    // console.log('sensitivity:          ', sensitivity);
+  }
+
+  public getBoundingBoxDiagonal(): number {
+    return getDiagonal(this._currentBoundingBox);
+  }
+  public getBoundingBoxDiagonal2(): number {
+    return getDiagonal2(this._currentBoundingBox);
+  }
+}
+
+function getDiagonal(boundingBox: Box3): number {
+  return boundingBox.min.distanceTo(boundingBox.max);
+}
+function getDiagonal2(boundingBox: Box3): number {
+  return Math.sqrt((boundingBox.min.x - boundingBox.max.x) ** 2 + (boundingBox.min.y - boundingBox.max.y) ** 2);
+}
+
+/**
+ * Finds 0, 1 or 2 intersection points, return number of intersections and
+ * sets the closest and furthest intersection points.
+ */
+function intersectBox(ray: Ray, box: Box3, closestIntersection: Vector3, furthestIntersection: Vector3): number {
+  const invdirx = 1 / ray.direction.x;
+  const invdiry = 1 / ray.direction.y;
+  const invdirz = 1 / ray.direction.z;
+
+  const origin = ray.origin;
+  let tmin, tmax;
+  if (invdirx >= 0) {
+    tmin = (box.min.x - origin.x) * invdirx;
+    tmax = (box.max.x - origin.x) * invdirx;
+  } else {
+    tmin = (box.max.x - origin.x) * invdirx;
+    tmax = (box.min.x - origin.x) * invdirx;
+  }
+  let tymin, tymax;
+  if (invdiry >= 0) {
+    tymin = (box.min.y - origin.y) * invdiry;
+    tymax = (box.max.y - origin.y) * invdiry;
+  } else {
+    tymin = (box.max.y - origin.y) * invdiry;
+    tymax = (box.min.y - origin.y) * invdiry;
+  }
+  if (tmin > tymax || tymin > tmax) return 0;
+  if (tymin > tmin || isNaN(tmin)) tmin = tymin;
+  if (tymax < tmax || isNaN(tmax)) tmax = tymax;
+
+  let tzmin, tzmax;
+  if (invdirz >= 0) {
+    tzmin = (box.min.z - origin.z) * invdirz;
+    tzmax = (box.max.z - origin.z) * invdirz;
+  } else {
+    tzmin = (box.max.z - origin.z) * invdirz;
+    tzmax = (box.min.z - origin.z) * invdirz;
+  }
+
+  if (tmin > tzmax || tzmin > tmax) return 0;
+  if (tzmin > tmin || tmin !== tmin) tmin = tzmin;
+  if (tzmax < tmax || tmax !== tmax) tmax = tzmax;
+
+  // return point closest to the ray (positive side)
+  if (tmin > tmax) [tmin, tmax] = [tmax, tmin];
```

Comment:
Is this necessary? From what I can see, `tmax` always seems to be larger than `tmin`. But not a big issue, anyway

Reply:
No :see_no_evil: